MachineSSAUpdater: use all vreg attributes instead of reg class only

petar-avramovic opened the pull request:
When initializing MachineSSAUpdater save all attributes of current
virtual register and create new virtual registers with same attributes.
Now new virtual registers have same both register class or bank and LLT.
Previously new virtual registers had same register class but LLT was not
set (LLT was set to default/empty LLT).
Required by GlobalISel for AMDGPU, new 'lane mask' virtual registers
created by MachineSSAUpdater need to have both register class and LLT.

patch 4 from: https://github.com/llvm/llvm-project/pull/73337

arsenm commented on llvm/include/llvm/CodeGen/MachineSSAUpdater.h:
```diff
@@ -40,8 +40,8 @@ class MachineSSAUpdater {
-  /// VRC - Register class of the current virtual register.
-  const TargetRegisterClass *VRC = nullptr;
+  /// RegAttrs - current virtual register, new registers copy its attributes.
+  Register RegAttrs;
```
Relying on an existing register (which may mutate?) may be cumbersome. Maybe would be better to store the RegClassOrRegBank + LLT?

petar-avramovic replied:
Proposed changes are in https://github.com/llvm/llvm-project/pull/73337, had to update amdgpu to do the same

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/divergence-divergent-i1-phis-no-lane-mask-merging.ll:
```diff
@@ -66,15 +65,16 @@ exit:
-; GFX10-NEXT:  .LBB1_3: ; %exit
-; GFX10-NEXT:    v_cndmask_b32_e64 v2, 0, -1, s0
+; GFX10-NEXT:    s_cbranch_scc1 .LBB1_2
+; GFX10-NEXT:  ; %bb.1: ; %B
```
code changes seem incidental from not running the test before?

petar-avramovic replied:
Original is from is patch 2 (bad code) + do-not-run-me from patch 3 because compilation fails.
This is patch 3 + patch 4 that now compiles successfully. The "andn2, and, or" part is from patch 3.

arsenm commented on llvm/lib/CodeGen/MachineSSAUpdater.cpp:
```diff
@@ -158,9 +153,9 @@ Register MachineSSAUpdater::GetValueInMiddleOfBlock(MachineBasicBlock *BB,
-                                        BB, BB->getFirstTerminator(),
-                                        VRC, MRI, TII);
+    MachineInstr *NewDef =
+        InsertNewDef(TargetOpcode::IMPLICIT_DEF, BB, BB->getFirstTerminator(),
```
If it's a generic virtual register, should switch to G_IMPLICIT_DEF/G_PHI

arsenm reviewed (changes requested):
Seems out of date, and needs to swap in G_* opcodes for generic registers

petar-avramovic committed:
MachineSSAUpdater: use all vreg attributes instead of reg class only

When initializing MachineSSAUpdater save all attributes of current
virtual register and create new virtual registers with same attributes.
Now new virtual registers have same both register class or bank and LLT.
Previously new virtual registers had same register class but LLT was not
set (LLT was set to default/empty LLT).
Required by GlobalISel for AMDGPU, new 'lane mask' virtual registers
created by MachineSSAUpdater need to have both register class and LLT.

patch 4 from: https://github.com/llvm/llvm-project/pull/73337

petar-avramovic commented:
> Seems out of date, and needs to swap in G_* opcodes for generic registers


For global-isel lane-mask registers are not generic registers, they have regclass + LLT instead of only regclass.
PHI lane masks should be fully instruction selected (all reg operands need to have reg class) when they are introduced.
Undef lane-masks should also be inst-selected, so use of IMPLICIT_DEF is fine.

arsenm commented:
> > Seems out of date, and needs to swap in G_* opcodes for generic registers
> 
> For global-isel lane-mask registers are not generic registers, they have regclass + LLT instead of only regclass. PHI lane masks should be fully instruction selected (all reg operands need to have reg class) when they are introduced. Undef lane-masks should also be inst-selected, so use of IMPLICIT_DEF is fine.

If that's the case, then this patch is unnecessary? Allowing MachineSSAUpdater in GlobalISel means handling generic virtual registers, and emitting the appropriate G_* opcodes instead of the post-selection equivalents.

arsenm commented:
> > > Seems out of date, and needs to swap in G_* opcodes for generic registers
> > 
> > 
> > For global-isel lane-mask registers are not generic registers, they have regclass + LLT instead of only regclass. PHI lane masks should be fully instruction selected (all reg operands need to have reg class) when they are introduced. Undef lane-masks should also be inst-selected, so use of IMPLICIT_DEF is fine.
> 
> If that's the case, then this patch is unnecessary? Allowing MachineSSAUpdater in GlobalISel means handling generic virtual registers, and emitting the appropriate G_* opcodes instead of the post-selection equivalents.

I guess preserve-LLT is one property, but is much more narrow than the title suggests. The title implies it will handle emitting G_* opcodes

petar-avramovic commented:
Rebase and ping, previously updated commit summary to not mention adapt for global-isel
